google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/DirectedReadTest.java

```diff
@@ -0,0 +1,251 @@
+                String.format(
+                    "set %sdirected_read='%s'",
+                    getVariablePrefix(), DirectedReadOptionsUtil.toString(expected))));
+        executeQuery(connection);
+        assertDirectedReadOptions(expected);
+
+        mockSpanner.clearRequests();
+
+        // Reset to default.
+        connection.execute(
+            Statement.of(String.format("set %sdirected_read=''", getVariablePrefix())));
+        executeQuery(connection);
+        assertDirectedReadOptions(DirectedReadOptions.getDefaultInstance());
+
+        mockSpanner.clearRequests();
+      }
+    }
+  }
+
+  @Test
+  public void testDirectedReadIsIgnoredInReadWriteTransaction() {
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      connection.setReadOnly(false);
+
+      connection.execute(
+          Statement.of(
+              String.format(
+                  "set %sdirected_read='%s'",
+                  getVariablePrefix(),
+                  DirectedReadOptionsUtil.toString(
+                      DirectedReadOptions.newBuilder()
+                          .setIncludeReplicas(
+                              IncludeReplicas.newBuilder()
+                                  .addReplicaSelections(
+                                      ReplicaSelection.newBuilder()
+                                          .setType(ReplicaSelection.Type.READ_WRITE)
+                                          .setLocation("us-west1")
+                                          .build())
+                                  .build())
+                          .build()))));
+      // This uses a read/write transaction, which will ignore any DirectedReadOptions.
+      executeQuery(connection);
+      assertDirectedReadOptions(DirectedReadOptions.getDefaultInstance());
+    }
+  }
+
+  @Test
+  public void testDirectedReadIsUsedInReadOnlyTransaction() {
+    DirectedReadOptions expected =
+        DirectedReadOptions.newBuilder()
+            .setIncludeReplicas(
+                IncludeReplicas.newBuilder()
+                    .addReplicaSelections(
+                        ReplicaSelection.newBuilder()
+                            .setType(ReplicaSelection.Type.READ_WRITE)
+                            .setLocation("us-west1")
+                            .build())
+                    .build())
+            .build();
+
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      connection.setReadOnly(true);
+
+      connection.execute(
+          Statement.of(
+              String.format(
+                  "set %sdirected_read='%s'",
+                  getVariablePrefix(), DirectedReadOptionsUtil.toString(expected))));
+      // This uses a read/write transaction, which will ignore any DirectedReadOptions.
+      executeQuery(connection);
+      assertDirectedReadOptions(expected);
+    }
+  }
+
+  private void executeQuery(Connection connection) {
```

Comment:
Would it be better to name it executeReadQuery?

Reply:
I'm not sure why you feel that. It calls the method `executeQuery`. Or am I missing something here?

If any, I would maybe otherwise suggest `executeQueryAndConsumeResults` (the latter is why I put it in a separate method, so we didn't have to repeat the empty `while(resultSet.next())` block in every test method.

Reply:
I've updated this method and added two more methods for executing both a read query and a DML query to ensure that we do the right thing in both cases.

(The original implementation did not do the right thing, as it would also include DirectedReadOptions for DML statements in auto-commit mode)